asciidoctor: include the asciidoctor-epub3 gem

michaelpj opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change
This is another common output format for asciidoctor, and this gem is at least semi-official (maintained by the asciidoctor maintainer).

###### Things done

The first commit does some cleaning, by allowing us to use the default gem config in nixpkgs. We need this in the second commit as it makes nokogiri work. We could combine the configs properly, but it seems better to put our configuration for the mathematical gem centrally in case other people need to use it.

I also had to change how the update script works, but I think it's reasonable still.

The second commit actually adds the `asciidoctor-epub3` gem. I've tested that this does actually work thereafter.

cc @worldofpeace @danbst who've looked at asciidoctor changes recently.

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

michaelpj committed:
asciidoctor: move mathematical gem config to default gemset config

This has two benefits:
- Anyone else using the `mathematical` gem will benefit from the config.
- The gem config overrides the default one, so we were losing the fixes
for other gems.

I had to change how the update script works. Now it looks at the `gems`
passthru from `bundlerApp`.

michaelpj committed:
asciidctor: include the asciidoctor-epub3 gem

This is another common output target for asciidctor that requires an
additional gem.

The previous commit is necessary for this to work, so that we get the
gem config for nokogiri correctly.

michaelpj commented on pkgs/tools/typesetting/asciidoctor/update.sh:
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env bash
-rm gemset.nix Gemfile.lock
```
If we do this before building the shell then we can't build the shell!

zimbatm replied:
Agreed! The more general issue is that the updater script is depending on the project to be working. So as soon as it breaks it becomes harder to repair it.

michaelpj replied:
I don't see a way to avoid this while still piggybacking off the app definition to get all the dependencies....

danbst replied:
:man_facepalming: probably I was running a nix-shell inside nix-shell... I wish there was `bundix --force-update`

zimbatm commented on pkgs/tools/typesetting/asciidoctor/default.nix:
```diff
@@ -1,59 +1,29 @@
-, cmake, bison, flex, glib, pkgconfig, cairo
-, pango, gdk_pixbuf, libxml2, python3, patchelf
-}:
+{ lib, bundlerApp, mkShell, bundix, defaultGemConfig }:
```
```suggestion
{ lib, bundlerApp, mkShell, bundix }:
```
defaultGemConfig doesn't seem to be used

zimbatm commented on pkgs/tools/typesetting/asciidoctor/default.nix:
```diff
@@ -1,59 +1,29 @@
+      [ bundix ]; 
   };
-}
+in app.overrideAttrs (attrs: { passthru = attrs.passthru // { updateShell = shell; }; })
```
```suggestion
in app
```
attach the shell directly on the app.passthru to avoid the overhead

michaelpj replied:
This doesn't work as `app.gems` isn't defined then.

zimbatm replied:
I pushed f726160556e74c8203f3f8dff0aa0dec07c0218c which demonstrates how to implement it. The key is that nix is a lazily-evaluated language.

zimbatm commented on pkgs/tools/typesetting/asciidoctor/default.nix:
```diff
@@ -1,59 +1,29 @@
+    buildInputs = 
+      app.gems.mathematical.buildInputs ++ 
+      app.gems.nokogiri.buildInputs ++ 
+      [ bundix ]; 
```
A more generic version would be:
```
inputsFrom = lib.attrValues app.gems;
buildInputs = [ bundix ];
```

michaelpj commented:
Also I've realised that I may want to include the `epubcheck` and `kindlegen` gems, since they don't seem to be properly included by default. Some bundler fanciness in the asciidoctor Gemfile. Investigating....

michaelpj committed:
asciidoctor: address review comments

michaelpj commented:
Comments addressed.

I've opted not to include the `epubcheck` and `kindlegen` dependencies. They're a bit dodgy, e.g. kindlegen just curls the binary from the internet and provides it. Even better, the paths to the epubcheck and kindlegen tools can just be given by environment variables, so there's no need to use the dodgy gems.

michaelpj commented:
Actually, given the environment variables work, I can just wrap the executable. The remaining annoyance is that kindlegen is unfree. I guess I could gate that behind `withKindlegen`? What's the usual approach for optional unfree deps?

michaelpj committed:
asciidoctor: wrap asciidoctor-epub3 for epubcheck and kindlegen

kindlegen is unfree, so we don't include it by default. The derivation
can be overridden to include it.

michaelpj commented:
Okay, I wrapped the executable and that works nicely. I've put something in so `epubcheck` and `kindlegen` are optional, with `kindlegen` being off by default due to being unfree. But if there's a more standard way, happy to use that.

zimbatm committed:
asciidoctor: fix update.sh

michaelpj commented:
Nice, thanks.

zimbatm commented:
@michaelpj let me know if you agree with my changes, if yes I will merge the PR.

michaelpj commented:
Yes, they're good!

lopsided98 commented:
This had the effect of breaking the default NixOS configuration on armv6l and armv7l, because it pulls Java into the build-time closure through epubcheck. Java has been broken on these platforms for quite a while, and this PR simply made the issue obvious.

On the other hand, this is the only usage of Java in a basic NixOS system, and pulls in many other dependencies (at build time, so binary cache users are not affected), so I am wondering if we want to disable epubcheck by default. I can't decide whether this is a good idea, or if we are better having a more fully featured package. Is there any specific use-case that needs epubcheck?

michaelpj commented:
Hm, I'd assumed that Java would be an unproblematic package! I also didn't know the default configuration depended on this (why? the manual?).

I think disabling epubcheck would be fine. Alternatively, we could override it to disable epubcheck at the use site which is problematic, e.g. where the manual uses it if that's the problem.

lopsided98 commented:
The dependency chain is: `udisks -> libblockdev -> libndctl -> asciidoctor`

zimbatm commented:
@michaelpj: can you send a PR to disable epubcheck please?

michaelpj commented:
@zimbatm https://github.com/NixOS/nixpkgs/pull/65018

lopsided98 commented:
I'm working on fixing Java on ARM as well, and should have a PR by this weekend.
